### ISSUE TITLE
Existing RedisInstances failing to update state due hard requirement on readReplicaMode having value

### ISSUE THREAD

**Issue by n1koo** [GoogleCloudPlatform/k8s-config-connector#622]
## Checklist

- [x] I did not find a related open issue.
- [ ] I did not find a solution in the troubleshooting guide: (https://cloud.google.com/config-connector/docs/troubleshooting)
- [ ] If this issue is time-sensitive, I have submitted a corresponding issue with [GCP support](https://cloud.google.com/support-hub).

## Bug Description

After upgrading to `1.7.4` from `1.6.1` all our old Redis instances started showing `'Update call failed: cannot make changes to immutable field(s): [readReplicasMode]'`

This is due to the instance being created before this feature was a thing. It seems its `NA` for old instances and cannot be changed.

**Comment by maqiuyujoyce** [comment]
Hi @n1koo , thank you for reporting the issue! I tried to create a RedisInstance in Config Connector `1.61.0` and then upgraded the version to `1.74.0`, but didn't encounter the error you described (see my yaml snippet and steps in the end of the comment). Could you share more details about this bug following the template below?

````
## Additional Diagnostic Information

### Kubernetes Cluster Version

<!--
Run the following command to get the Kubernetes cluster version:

kubectl version --short
-->

### Config Connector Version

<!--
Run the following command to get the installed Config Connector version:

kubectl get ns cnrm-system -o jsonpath='{.metadata.annotations.cnrm\.cloud\.google\.com/version}'
-->

### Config Connector Mode

<!--
Run the following command to get the Config Connector Mode:

kubectl get ConfigConnector "configconnector.core.cnrm.cloud.google.com" -o=jsonpath="{@.spec.mode}"
-->

## Log Output

<!--
Include any relevant log output. See
  https://cloud.google.com/config-connector/docs/troubleshooting#logging

If the issue is for a specific resource, please include those logs.
-->

## Steps to Reproduce

### Steps to reproduce the issue
<!--- Steps needed to reproduce the issue. --->

### YAML snippets
<!--- YAML snippets needed to reproduce the issue. See the following as an example. --->

```yaml
apiVersion: pubsub.cnrm.cloud.google.com/v1beta1
kind: PubSubTopic
metadata:
  labels:
    label-one: "value-one"
  name: pubsubtopic-sample
```
````

====== YAML and steps I tried ======
The YAML snippet I used:
```
apiVersion: redis.cnrm.cloud.google.com/v1beta1
kind: RedisInstance
metadata:
  name: redisinstance-test
  namespace: [namespace]
spec:
  displayName: Sample Redis Instance
  region: us-central1
  tier: BASIC
  memorySizeGb: 1
```

Steps:
1. Installed Config Connector `1.61.0` in cluster mode.
2. Created a `RedisInstance` using the YAML snippet above and waited for for it to be ready.
3. Upgraded Config Connector to `1.74.0`.
4. Run `kubectl describe` to check the status of the RedisInstance and it is `UpToDate`.

**Comment by n1koo** [comment]
@maqiuyujoyce the problem exists with older redis instances. In our case they were atleast 270days old.

Otherwise your steps are correct, but you require an older instance to reproduce.


For example comparing a new and old instance via `gcloud`:

Old (as in older than 9months)
```
‽ gcloud redis instances describe <new-instance>  --region <region> --project <project>|grep read
(nothing matched)
```

In console showing `Read Replicas: NA`

New(as in newer than 9months)
```
‽ gcloud redis instances describe <new-instance>  --region <region> --project <project>|grep read
readReplicasMode: READ_REPLICAS_DISABLED
```

**Comment by maqiuyujoyce** [comment]
Hi @n1koo , thanks for sharing the context.

Have you tried to [abandon](https://cloud.google.com/config-connector/docs/how-to/managing-deleting-resources#keeping_resources_after_deletion) the RedisInstance and acquire it with `readReplicasMode` explicitly set to `NA`?

**Comment by n1koo** [comment]
Sorry @maqiuyujoyce not yet. We reverted the upgrade for now. Doing that dance for our instances would be somewhat painful though as its not just a handful.

We could write that automation, but I was a little bit hoping config-connector could handle it better internally? 

This change is not mentioned even as breaking change

**Comment by jcanseco** [comment]
Hi @n1koo, this is indeed strange behavior.

Can you get the state of your Redis instances by hitting the API directly? Use "Try It!" here: https://cloud.google.com/memorystore/docs/redis/reference/rest/v1/projects.locations.instances/get

I'm particularly interested in the `readReplicasMode` value returned by the API.

Also: can you verify that your `RedisInstance` resources don't specify a `readReplicasMode` after the KCC upgrade?

**Comment by jcanseco** [comment]
It may be worth noting that "NA" is _not_ a valid `readReplicasMode` value, but this should be irrelevant as long as there is no diff between the resource configuration that KCC tries to apply and the resource state that KCC reads from the API.

As far as I can tell, we are not modifying the `readReplicasMode` value we read back from the API at all even if the value is technically invalid.

Unfortunately, this is also going to be a bit challenging to debug for us since we do not have "old" Redis instances to work with.

**Comment by jcanseco** [comment]
Ok, I suspect the API is actually returning `readReplicasMode: ""` -- can you confirm this @n1koo?

**Comment by jcanseco** [comment]
Hi @n1koo. Gotcha, I'll check in with the Redis team internally, but if you can, it would really help us if you can confirm what value for `readReplicasMode` is returned by the API. Checking what a client library returns is insufficient unfortunately since they don't necessarily fully reflect what the API actually returns.

**Comment by jcanseco** [comment]
Hi @n1koo, is this something you can report to [GCP support ](https://cloud.google.com/support-hub)? 

I chatted with the Redis team internally and it seems that old Redis instances should still also have a value for `readReplicasMode` (`READ_REPLICAS_DISABLED`).

If you are seeing that your instances are not reporting a value for `readReplciasMode` (even if it is through `google-cloud-redis`), I think that might be worth reporting since that appears to be unintended behavior.

**Comment by jcanseco** [comment]
Nevermind! It seems Terraform users have also reported facing this issue: https://github.com/hashicorp/terraform-provider-google/issues/11410

And Terraform team _is_ able to reproduce the issue as they have an old Redis instance that predates the read replicas feature. Let me work with the Terraform and Redis teams internally to see what we can do.

**Comment by n1koo** [comment]
Awesome thank you! I also finally got the access so heres the raw api response;


```
{
  "name": "projects/xxx/locations/xxx/instances/xxx",
  "displayName": "xx",
  "labels": {
    //xx
  },
  "locationId": "us-east1-b",
  "redisVersion": "REDIS_4_0",
  "reservedIpRange": "xxx",
  "host": "xxx",
  "port": 6379,
  "currentLocationId": "us-east1-b",
  "createTime": "2020-04-22T19:51:04.021313435Z",
  "state": "READY",
  "tier": "BASIC",
  "memorySizeGb": 8,
  "authorizedNetwork": "projects/xxx/global/networks/xxx",
  "persistenceIamIdentity": "serviceAccount:xxx-compute@developer.gserviceaccount.com",
  "connectMode": "DIRECT_PEERING",
  "transitEncryptionMode": "DISABLED",
  "maintenancePolicy": {
    "createTime": "2021-08-07T11:01:09.836707Z",
    "updateTime": "2021-08-07T11:01:09.836707Z",
    "weeklyMaintenanceWindow": [
      {
        "day": "SATURDAY",
        "startTime": {
          "hours": 5
        },
        "duration": "3600s"
      }
    ]
  },
  "nodes": [
    {
      "id": "node-0",
      "zone": "us-east1-b"
    }
  ]
}
```

**Comment by jcanseco** [comment]
Yep that is correct. We'll post an update once this has been fixed :)

**Comment by jcanseco** [comment]
@Veres72 thank you for sharing! Yes that workaround makes sense and is what we recommend for now.

We are looking to release a new Config Connector version with a fix which we intend to also roll out to GKE 1.23+ clusters using the KCC add-on.

We are also asking the Redis team to implement a server-side fix to accommodate users who may not be able to upgrade for whatever reason, but we have not received updates from them yet.
